kmir/src/kmir/kdist/mir-semantics/rt/data.md

```diff
@@ -870,6 +929,68 @@ which have the same representation `Value::Range`.
        andBool isTypeInfo(TYPEMAP[TY_TARGET])
        andBool #typesCompatible({TYPEMAP[TY_SOURCE]}:>TypeInfo, {TYPEMAP[TY_TARGET]}:>TypeInfo, TYPEMAP)
       [preserves-definedness] // valid map lookups checked
+
+  syntax PtrEmulation ::= #convertPtrEmul ( PtrEmulation , TypeInfo , Map ) [function, total]
+  // ----------------------------------------------------------------------------------
+```
+
+Pointers to slices can be converted to pointers to single elements, _losing_ their metadata.
+```k
+  rule #convertPtrEmul(     ptrEmulation(_)           , typeInfoRefType(POINTEE_TY), TYPEMAP) => ptrEmulation(noMetadata)
+    requires #metadata(POINTEE_TY, TYPEMAP) ==K noMetadata                                                                [priority(60)]
+  rule #convertPtrEmul(     ptrEmulation(_)           , typeInfoPtrType(POINTEE_TY), TYPEMAP) => ptrEmulation(noMetadata)
+    requires #metadata(POINTEE_TY, TYPEMAP) ==K noMetadata                                                                [priority(60)]
+```
+
+Conversely, when casting a pointer to an element to a pointer to a slice or array,
+the original allocation size must be checked to be sufficient.
+
+**TODO** this is future work, requiring a "provenance" that retains the original allocation size. Invalid casts should be marked with special "InvalidCast" metadata.
+
+```k
+  // no metadata to begin with, fill it in from target type (NB dynamicSize(1) if dynamic)
+  rule #convertPtrEmul(   ptrEmulation(noMetadata)    , typeInfoRefType(POINTEE_TY), TYPEMAP) => ptrEmulation(#metadata(POINTEE_TY, TYPEMAP))
+  rule #convertPtrEmul(   ptrEmulation(noMetadata)    , typeInfoPtrType(POINTEE_TY), TYPEMAP) => ptrEmulation(#metadata(POINTEE_TY, TYPEMAP))
+```
```

Comment:
In the next step, we would want to retain the original allocation for pointers, in order to remember whether it is OK to convert to a certain (static) array size later. 
The example program is converting a `*[u8]` through `*u8` to `[u8; 9]` (undefined behaviour), the original `*[u8]` having a known dynamic size means we could detect it. However, the `Deref` access will anyway fail later, we could just mark the case as invalid metadata ahead of time.